app/channel_share.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/store"
+)
+
+// SharedChannels
+
+func (a *App) SaveSharedChannel(sc *model.SharedChannel) (*model.SharedChannel, error) {
+	// Check channel is not already shared.
+	if _, err := a.GetSharedChannel(sc.ChannelId); err == nil {
+		return nil, errors.New("channel is already shared.")
+	}
+
+	// check that channel exists.
+	if _, errApp := a.GetChannel(sc.ChannelId); errApp != nil {
+		return nil, fmt.Errorf("Cannot share this channel: %v", errApp)
+	}
```

Comment:
Thanks for reviewing @isacikgoz.  Capital is fixed.  I'm not clear on what you mean by "prior check".  I can't do this further up the call stack since slash command, rest API, services can all consume this API and we don't want to repeat the check everywhere.

Reply:
I mean, we first check if the channel is shared then check its existence. Shouldn't we check if the channel exists first?

Reply:
Ah I see. Yes, I can do that.  FYI - checking the channel is shared is a different table.  So for shared=true and GetChannel to fail would be a database connection error or the shared channel record became orphaned (yikes!).  That said changing the order is fine.

Reply:
Done